### PR TITLE
JAVA-2715: Query builder for new lookup stage

### PR DESCRIPTION
**Pull request by fhassak**
thanks to @csarrazi who helped me for this contribution

### CODE REVIEW

**Comment by csarrazi** [comment]
You’re welcome! 😉

File: driver-core/src/main/com/mongodb/client/model/Aggregates.java [hunk 1]
```diff
@@ -243,6 +243,38 @@ public static Bson lookup(final String from, final String localField, final Stri
...
+     * @param from          the name of the collection in the same database to perform the join with.
+     * @param pipeline      the pipeline to run on the joined collection.
+     * @param as            the name of the new array field to add to the input documents.
+     * @return the $lookup pipeline stage
+     * @mongodb.driver.manual reference/operator/aggregation/lookup/ $lookup
+     * @mongodb.server.release 3.6
+     * @since 3.7
+     *
+     */
+    public static Bson lookup(final String from, final List<? extends Bson> pipeline, final String as) {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/client/model/Aggregates.java]: I'd like to see a unit and functional test for this overload.  Currently it's untested, but I want to make sure we catch any issues when `let` is null.

File: driver-core/src/main/com/mongodb/client/model/Aggregates.java [hunk 2]
```diff
@@ -243,6 +243,38 @@ public static Bson lookup(final String from, final String localField, final Stri
                                                    .append("as", new BsonString(as)));
     }
 
+    /**
+     * Creates a $lookup pipeline stage for the specified filter
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/client/model/Aggregates.java]: Let's fix this comment now, as there is no `filter`, but rather a pipeline.  Perhaps:

   Creates a $lookup pipeline stage, joining the current collection with the one specified in `from` using the given `pipeline`.

File: driver-core/src/main/com/mongodb/client/model/Aggregates.java [hunk 3]
```diff
@@ -243,6 +243,38 @@ public static Bson lookup(final String from, final String localField, final Stri
...
+     */
+    public static Bson lookup(final String from, final List<? extends Bson> pipeline, final String as) {
+        return lookup(from, null, pipeline, as);
+    }
+
+    /**
+     * Creates a $lookup pipeline stage for the specified filter
+     *
+     * @param from          the name of the collection in the same database to perform the join with.
+     * @param let           specifies variables to use in the pipeline field stages.
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/client/model/Aggregates.java]: Replace "specifies" with "the".

File: driver-core/src/main/com/mongodb/client/model/Aggregates.java (hunk 2 above)
Comment (jyemin) [reply driver-core/src/main/com/mongodb/client/model/Aggregates.java]: (I realize the comment in the existing lookup method is misleading as well.  Feel free to fix that one too if you're so inclined.)

**Comment by fhassak** [comment]
It's hard for me to know what's wrong because it works on my pc :

image

File: driver-core/src/test/functional/com/mongodb/client/model/AggregatesFunctionalSpecification.groovy [hunk 4]
```diff
@@ -291,6 +291,60 @@ class AggregatesFunctionalSpecification extends OperationFunctionalSpecification
...
+        fromCollection.insertDocuments(
+                Document.parse('{ "_id" : 1, year: 2018, name: "New Years", date: { $date : "2018-01-01T00:00:00Z"} }'),
+                Document.parse('{ "_id" : 2, year: 2018, name: "Pi Day", date: { $date : "2018-03-14T00:00:00Z" } }'),
+                Document.parse('{ "_id" : 3, year: 2018, name: "Ice Cream Day", date: { $date : "2018-07-15T00:00:00Z"} }'),
+                Document.parse('{ "_id" : 4, year: 2017, name: "New Years", date: { $date : "2017-01-01T00:00:00Z" } }'),
+                Document.parse('{ "_id" : 5, year: 2017, name: "Ice Cream Day", date: { $date : "2017-07-16T00:00:00Z" } }')
+        )
+
+        collection.insertDocuments(
+                Document.parse('''{ "_id" : 1, "student" : "Ann Aardvark",
```
Comment (jyemin) [inline driver-core/src/test/functional/com/mongodb/client/model/AggregatesFunctionalSpecification.groovy]: The build failure is due to trailing whitespace on this line.  If you run  `./gradlew driver-core:check' it should fail for you too.

**Comment by jyemin** [comment]
Thanks for the work on this @fhassak!  It's now merged to master.

**Pull request merged**